### PR TITLE
refactor: handle viem chains in wagmi adapter constructor

### PR DESCRIPTION
**Pull request by enesozturk**
# Description

The current Wagmi adapter doesn't handle the Viem chains well. Most of the users use `viem`'s [exported chains](https://github.com/wevm/viem/tree/main/src/chains/definitions). The Wagmi adapter should be flexible to use all these chains as well.

This PR introduces changes to WagmiAdapter and lab examples to provide proper workaround handling both Viem chains and our exported network objects.

## Type of change

- [x] Chore (non-breaking change that addresses non-functional tasks, maintenance, or code quality improvements)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

# Associated Issues

For Linear issues: Closes APKT-1193, APKT-1194, APKT-1195
For GH issues: closes #...

# Showcase (Optional)

If there is a UI change include the screenshots with before and after state.
If new feature is being introduced, include the link to demo recording.

# Checklist

- [x] Code in this PR is covered by automated tests (Unit tests, E2E tests)
- [x] My changes generate no new warnings
- [x] I have reviewed my own code
- [x] I have filled out all required sections
- [x] I have tested my changes on the preview link
- [ ] Approver of this PR confirms that the changes are tested on the preview link

### CODE REVIEW

File: apps/laboratory/src/components/UPA/UpaSolanaSignMessageTest.tsx [hunk 1]
```diff
@@ -25,7 +25,7 @@ export function UpaSolanaSignMessageTest() {
           pubkey: address
         }
       }
-      const chain = caipNetwork?.id
+      const chain = caipNetwork?.id?.toString()
```
Comment (enesozturk) [inline apps/laboratory/src/components/UPA/UpaSolanaSignMessageTest.tsx:28]: Why do we have to do this?

File: apps/laboratory/src/pages/library/wagmi.tsx [hunk 2]
```diff
@@ -18,8 +18,12 @@ const wagmiAdapter = new WagmiAdapter({
 
 const modal = createAppKit({
   adapters: [wagmiAdapter],
-  networks: ConstantsUtil.EvmNetworks,
+  networks: wagmiAdapter.caipNetworks,
   projectId: ConstantsUtil.ProjectId,
+  chainImages: {
```
Comment (enesozturk) [inline apps/laboratory/src/pages/library/wagmi.tsx:23]: We could remove. I've added this to prove Viem chains could be used but not they are available out of the box

File: packages/adapters/ethers/src/tests/client.test.ts [hunk 3]
```diff
@@ -26,13 +37,16 @@ vi.mock('@reown/appkit-wallet', () => ({
   }
 }))
 
-vi.mock('@reown/appkit-utils', () => {
+vi.mock('@reown/appkit-utils', async importOriginal => {
+  const actual = await importOriginal()
```
Comment (enesozturk) [inline packages/adapters/ethers/src/tests/client.test.ts:41]: To not mock `CaipNetworkUtil` to extend networks as expected on test environment

File: packages/adapters/ethers/src/tests/client.test.ts [hunk 4]
```diff
@@ -144,16 +158,9 @@ describe('EthersAdapter', () => {
     })
 
     it('should switch network for injected provider', async () => {
-      const newNetwork = {
```
Comment (enesozturk) [inline packages/adapters/ethers/src/tests/client.test.ts:147]: We don't have to mock networks, use directly from exported networks

File: packages/adapters/solana/src/tests/client.test.ts [hunk 5]
```diff
@@ -65,7 +65,7 @@ describe('SolanaAdapter', () => {
   beforeEach(() => {
     vi.clearAllMocks()
     client = new SolanaAdapter({})
-    client.construct(mockAppKit, mockOptions)
+    client.construct(mockAppKit, mockOptions, [solana])
```
Comment (enesozturk) [inline packages/adapters/solana/src/tests/client.test.ts:68]: Does the `construct` method have to get a third parameter?

File: packages/adapters/wagmi/src/client.ts [hunk 6]
```diff
@@ -138,28 +145,26 @@ export class WagmiAdapter implements ChainAdapter {
...
-      )
-    )
+    this.wagmiChains = this.caipNetworks.filter(
+      caipNetwork => caipNetwork.chainNamespace === CommonConstantsUtil.CHAIN.EVM
+    ) as unknown as [BaseNetwork, ...BaseNetwork[]]
 
     const transportsArr = this.wagmiChains.map(chain => [
       chain.id,
-      getTransport({ chain, projectId: configParams.projectId })
+      getTransport({ chain: chain as Chain, projectId: configParams.projectId })
```
Comment (enesozturk) [inline packages/adapters/wagmi/src/client.ts:167]: What is `Chain`? We might need to use types from common/TypeUtils

File: packages/adapters/wagmi/src/client.ts [hunk 7]
```diff
@@ -470,11 +488,12 @@ export class WagmiAdapter implements ChainAdapter {
               'networkControllerClient:getApprovedCaipNetworksData - wagmiConfig is undefined'
             )
           }
-          const chainId = Number(
-            NetworkUtil.caipNetworkIdToNumber(this.appKit?.getCaipNetwork()?.id)
-          )
+          const activeCaipNetwork = this.appKit?.getCaipNetwork()
+          const chainId = activeCaipNetwork?.id as number | undefined
```
Comment (enesozturk) [inline packages/adapters/wagmi/src/client.ts:492]: These are duplicationg, maybe we could add `getCaipNetwork` function?

File: packages/appkit-utils/src/CaipNetworkUtil.ts [hunk 8]
```diff
@@ -0,0 +1,170 @@
...
+} from '@reown/appkit-common'
+import { PresetsUtil } from './PresetsUtil.js'
+
+const RPC_URL_HOST = 'rpc.walletconnect.org'
+
+export function getBlockchainApiRpcUrl(caipNetworkId: CaipNetworkId, projectId: string) {
+  return `https://rpc.walletconnect.org/v1/?chainId=${caipNetworkId}&projectId=${projectId}`
+}
+
+const WC_HTTP_RPC_SUPPORTED_CHAINS = [
```
Comment (enesozturk) [inline packages/appkit-utils/src/CaipNetworkUtil.ts:15]: Maybe we could move to another place?

File: packages/appkit-utils/src/ethers/EthersHelpersUtil.ts [hunk 9]
```diff
@@ -1,22 +1,8 @@
-import { ConstantsUtil as CommonConstantsUtil } from '@reown/appkit-common'
 import type { CaipNetwork } from '@reown/appkit-common'
-import { ConstantsUtil } from '../ConstantsUtil.js'
 import { PresetsUtil } from '../PresetsUtil.js'
 import type { Provider } from './EthersTypesUtil.js'
 
 export const EthersHelpersUtil = {
-  getCaipDefaultChain(chain?: CaipNetwork) {
```
Comment (enesozturk) [inline packages/appkit-utils/src/ethers/EthersHelpersUtil.ts:8]: We use it nowhere

File: packages/appkit/package.json [hunk 10]
```diff
@@ -93,15 +93,18 @@
...
     "@reown/appkit-siwe": "workspace:*",
     "@reown/appkit-ui": "workspace:*",
+    "@reown/appkit-utils": "workspace:*",
     "@reown/appkit-wallet": "workspace:*",
     "@walletconnect/types": "2.17.0",
     "@walletconnect/utils": "2.17.0",
     "@walletconnect/universal-provider": "2.17.0",
     "valtio": "1.11.2"
   },
+  "peerDependencies": {
```
Comment (enesozturk) [inline packages/appkit/package.json:105]: Should this be `peerDependency`?

File: packages/appkit/src/client.ts [hunk 11]
```diff
@@ -512,24 +532,17 @@ export class AppKit {
     ChainController.initialize(options.adapters || [])
     options.adapters?.forEach(adapter => {
       // @ts-expect-error will introduce construct later
-      adapter.construct?.(this, options)
+      adapter.construct?.(this, options, this.caipNetworks)
```
Comment (enesozturk) [inline packages/appkit/src/client.ts:535]: Again, do we have to pass a third parameter?

File: packages/appkit/src/networks/eip155.ts [hunk 12]
```diff
@@ -0,0 +1,17 @@
+import type { CaipNetwork } from '@reown/appkit-common'
```
Comment (enesozturk) [inline packages/appkit/src/networks/eip155.ts:1]: Remove

File: packages/appkit/src/networks/index.ts [hunk 13]
```diff
@@ -1,195 +1,66 @@
...
 
-export const baseSepolia: CaipNetwork = {
-  id: 'eip155:84532',
-  chainId: 84532,
-  name: 'Base Sepolia',
-  currency: 'BASE',
-  explorerUrl: 'https://sepolia.basescan.org',
-  rpcUrl: getBlockchainApiRpcUrl(84532, 'eip155'),
-  chainNamespace: 'eip155'
+export function defineChain<
```
Comment (enesozturk) [inline packages/appkit/src/networks/index.ts:14]: Maybe move this function to another file and keep this file only for network objects?

File: packages/appkit/src/utils/HelpersUtil.ts [hunk 14]
```diff
@@ -50,15 +51,15 @@ export const WcHelpersUtil = {
         } satisfies Namespace
       }
 
-      const fullChainId = `${chainNamespace}:${chainId}`
+      const fullChainId = `${chainNamespace}:${id}`
```
Comment (enesozturk) [inline packages/appkit/src/utils/HelpersUtil.ts:54]: `caipNetworkId` would be better naming

File: packages/common/src/utils/TypeUtil.ts [hunk 15]
```diff
@@ -6,7 +31,7 @@ export type ChainId = string | number
 
 export type ChainNamespace = 'eip155' | 'solana' | 'polkadot'
 
-export type CaipNetwork = {
+export type CaipNetworkLegacy = {
```
Comment (enesozturk) [inline packages/common/src/utils/TypeUtil.ts:34]: If we don't use this anywhere, remove

File: packages/core/exports/vue.ts [hunk 16]
```diff
@@ -0,0 +1,58 @@
+import { AccountController } from '../src/controllers/AccountController.js'
+import { CoreHelperUtil } from '../src/utils/CoreHelperUtil.js'
+import { ChainController } from '../src/controllers/ChainController.js'
+import type { CaipNetwork, CaipNetworkId } from '@reown/appkit-common'
+import { ref } from 'valtio/vanilla'
+import { onUnmounted } from 'vue'
+
+// -- Hooks ------------------------------------------------------------
+export function useAppKitNetwork(): {
```
Comment (enesozturk) [inline packages/core/exports/vue.ts:9]: This might be a separate PR. We don't properly provide hooks for Vue. This is fixing that

File: packages/wagmi/package.json [hunk 17]
```diff
@@ -75,7 +75,7 @@
     "@wagmi/core": ">=2.0.0",
     "react": ">=17",
     "react-dom": ">=17",
-    "viem": ">=2.0.0",
+    "viem": "2.x",
```
Comment (enesozturk) [inline packages/wagmi/package.json:78]: Why there are two in this package.json?

File: apps/laboratory/src/components/UPA/UpaSolanaSignMessageTest.tsx (hunk 1 above)
Comment (enesozturk) [reply apps/laboratory/src/components/UPA/UpaSolanaSignMessageTest.tsx:28]: Oh this should be using `caipNetworkId` - fixed

File: packages/adapters/solana/src/tests/client.test.ts (hunk 5 above)
Comment (enesozturk) [reply packages/adapters/solana/src/tests/client.test.ts:68]: Resolved

File: apps/laboratory/src/components/Solana/SolanaTests.tsx [hunk 18]
```diff
@@ -99,8 +99,7 @@ export function SolanaTests() {
             <SolanaSignJupiterSwapTest />
           </Box>
 
-          {(caipNetwork?.chainId === solanaTestnet.chainId ||
-            caipNetwork?.chainId === solanaDevnet.chainId) && (
+          {(caipNetwork?.id === solanaTestnet.id || caipNetwork?.id === solanaDevnet.id) && (
```
Comment (tomiir) [inline apps/laboratory/src/components/Solana/SolanaTests.tsx:102]: is this ok?

File: apps/laboratory/src/pages/library/wagmi.tsx (hunk 2 above)
Comment (tomiir) [reply apps/laboratory/src/pages/library/wagmi.tsx:23]: let's remove 🙏

File: packages/adapters/ethers/src/client.ts [hunk 19]
```diff
@@ -828,8 +822,8 @@ export class EthersAdapter {
     }
 
     this.appKit?.setLoading(true)
-    const chainId = NetworkUtil.caipNetworkIdToNumber(this.appKit?.getCaipNetwork()?.id)
-    this.appKit?.setCaipAddress(`eip155:${chainId}:${address}`, this.chainNamespace)
+    const chainId = this.appKit?.getCaipNetwork()?.id
+    this.appKit?.setCaipAddress(`${this.chainNamespace}:${chainId}:${address}`, this.chainNamespace)
```
Comment (tomiir) [inline packages/adapters/ethers/src/client.ts:826]: is this change ok? won't it break anything?

File: packages/adapters/ethers/src/client.ts [hunk 20]
```diff
@@ -868,16 +862,19 @@ export class EthersAdapter {
     if (address) {
       if (isEipNetwork) {
         this.appKit?.setPreferredAccountType(preferredAccountType, this.chainNamespace)
-        this.appKit?.setCaipAddress(`${caipNetworkId}:${address}`, this.chainNamespace)
+        this.appKit?.setCaipAddress(
+          `${this.chainNamespace}:${caipNetworkId}:${address}`,
```
Comment (tomiir) [inline packages/adapters/ethers/src/client.ts:866]: you sure this doesn't break anything? On solana I believe we receive a `CaipNetworkId`as response, causing this to be like eip155:solana:andsasdad:...

File: packages/adapters/ethers5/src/client.ts [hunk 21]
```diff
@@ -828,8 +827,8 @@ export class Ethers5Adapter {
     }
 
     this.appKit?.setLoading(true)
-    const chainId = NetworkUtil.caipNetworkIdToNumber(this.appKit?.getCaipNetwork()?.id)
-    this.appKit?.setCaipAddress(`eip155:${chainId}:${address}`, this.chainNamespace)
+    const chainId = this.appKit?.getCaipNetwork()?.id
+    this.appKit?.setCaipAddress(`${this.chainNamespace}:${chainId}:${address}`, this.chainNamespace)
```
Comment (tomiir) [inline packages/adapters/ethers5/src/client.ts:831]: same as before

File: packages/adapters/solana/src/client.ts [hunk 22]
```diff
@@ -360,9 +362,14 @@ export class SolanaAdapter implements ChainAdapter {
...
   }) {
     if (address && caipNetwork) {
-      SolStoreUtil.setConnection(new Connection(caipNetwork.rpcUrl, this.connectionSettings))
+      SolStoreUtil.setConnection(
+        new Connection(caipNetwork.rpcUrls.default.http?.[0] as string, this.connectionSettings)
+      )
       this.appKit?.setAllAccounts([{ address, type: 'eoa' }], this.chainNamespace)
-      this.appKit?.setCaipAddress(`${caipNetwork.id}:${address}` as const, this.chainNamespace)
+      this.appKit?.setCaipAddress(
+        `${this.chainNamespace}:${caipNetwork.id}:${address}` as CaipAddress,
```
Comment (tomiir) [inline packages/adapters/solana/src/client.ts:370]: same

File: packages/adapters/wagmi/src/client.ts (hunk 6 above)
Comment (tomiir) [reply packages/adapters/wagmi/src/client.ts:167]: `Chain` is viem's chain type. 
If this is the same type for other networks we can keep, else we should replace for our own type

File: packages/adapters/wagmi/src/client.ts (hunk 7 above)
Comment (tomiir) [reply packages/adapters/wagmi/src/client.ts:492]: `getCaipNetworkId` in any case I think

File: packages/adapters/wagmi/src/tests/client.test.ts [hunk 23]
```diff
@@ -166,44 +172,42 @@ describe('Wagmi Client', () => {
...
       expect(setAddressExplorerUrlSpy).toHaveBeenCalledWith(
         'https://etherscan.io/address/0x1234567890123456789012345678901234567890',
         'eip155'
       )
-      expect(syncBalanceSpy).toHaveBeenCalledWith(mockAddress, mockChain.chainId)
+      expect(syncBalanceSpy).toHaveBeenCalledWith(mockAddress, mainnet.id)
     })
 
-    it('should not sync network if chain is not found', async () => {
+    it.skip('should not sync network if chain is not found', async () => {
```
[the remark this answers is not in the document]
Comment (tomiir) [reply packages/adapters/wagmi/src/tests/client.test.ts:204]: +1

File: packages/appkit-utils/src/CaipNetworkUtil.ts (hunk 8 above)
Comment (tomiir) [reply packages/appkit-utils/src/CaipNetworkUtil.ts:15]: do we want to keep and mantain this here? shouldn't we query this from an api call?

File: packages/appkit/package.json (hunk 10 above)
Comment (tomiir) [reply packages/appkit/package.json:105]: don't think that's valid 🤔

File: packages/appkit/src/networks/eip155.ts (hunk 12 above)
Comment (tomiir) [reply packages/appkit/src/networks/eip155.ts:1]: ?

File: packages/appkit/src/networks/index.ts (hunk 13 above)
Comment (tomiir) [reply packages/appkit/src/networks/index.ts:14]: I think it's fine here if not used elsewehere

File: packages/wagmi/package.json (hunk 17 above)
Comment (tomiir) [reply packages/wagmi/package.json:78]: is this ok? peer dependencies should be marked as a range of versions

File: packages/adapters/ethers/src/client.ts (hunk 19 above)
Comment (enesozturk) [reply packages/adapters/ethers/src/client.ts:826]: Nope, we should avoid hard coded strings

File: packages/adapters/ethers5/src/client.ts (hunk 21 above)
Comment (enesozturk) [reply packages/adapters/ethers5/src/client.ts:831]: Same, we should avoid hard coded strings

File: packages/appkit-utils/src/CaipNetworkUtil.ts (hunk 8 above)
Comment (enesozturk) [reply packages/appkit-utils/src/CaipNetworkUtil.ts:15]: I thought about it but I don't want to make the initialization slower. It'll add extra 200-300 ms time (maybe more depending on the internet)

File: apps/laboratory/src/utils/ConstantsUtil.ts [hunk 24]
```diff
@@ -58,22 +59,27 @@ const EvmNetworks = [
   zkSync,
   arbitrum,
   base,
+  baseSepolia,
   sepolia,
-  gnosis,
-  baseSepolia
-]
+  gnosis
+] as [AppKitNetwork, ...AppKitNetwork[]]
```
Comment (magiziz) [inline apps/laboratory/src/utils/ConstantsUtil.ts:65]: Can we do `as const` ?

File: apps/laboratory/src/utils/ConstantsUtil.ts [hunk 25]
```diff
@@ -58,22 +59,27 @@ const EvmNetworks = [
...
+    default: { name: 'Solscan', url: 'https://explorer.solana.com/?cluster=unsupported' }
+  },
+  rpcUrls: { default: { http: ['https://api.unsupported.solana.com'] } }
 } as CaipNetwork
 
-const SolanaNetworks = [solana, solanaTestnet, solanaDevnet, solanaNotExist]
+const SolanaNetworks = [solana, solanaTestnet, solanaDevnet, solanaNotExist] as [
+  AppKitNetwork,
+  ...AppKitNetwork[]
+]
```
Comment (magiziz) [inline apps/laboratory/src/utils/ConstantsUtil.ts:82]: Same here ?

File: apps/laboratory/src/utils/ConstantsUtil.ts [hunk 26]
```diff
@@ -131,5 +137,5 @@ export const ConstantsUtil = {
   ProjectId: projectId,
   EvmNetworks,
   SolanaNetworks,
-  AllNetworks: [...EvmNetworks, ...SolanaNetworks]
+  AllNetworks: [...EvmNetworks, ...SolanaNetworks] as [AppKitNetwork, ...AppKitNetwork[]]
```
Comment (magiziz) [inline apps/laboratory/src/utils/ConstantsUtil.ts:140]: Here as well ?

**Review by magiziz — APPROVED** [review]

File: apps/laboratory/src/utils/ConstantsUtil.ts (hunk 24 above)
Comment (enesozturk) [reply apps/laboratory/src/utils/ConstantsUtil.ts:65]: The `createAppKit`'s `networks` parameter accepts `[AppKitNetwork, ...AppKitNetwork[]]`. IMO it's better to make sure it's happy with the type definitions.
